File: solutions/TypeScript/[1288_Medium] Remove Covered Intervals/removeCoveredIntervals.ts
function insertionSort(arr: number[][]) {
  for (let k = 1; k < arr.length; k++) {
    const key: number[] = arr[k];
    let i: number = k - 1;
    while (i >= 0 && (arr[i][0] > key[0] || (arr[i][0] === key[0] && arr[i][1] < key[1]))) {
      arr[i + 1] = arr[i];
      i--;
    }
    arr[i + 1] = key;
  }
}

function removeCoveredIntervals(intervals: number[][]): number {
  insertionSort(intervals);

  let count = 1;
  let rangeMax = intervals[0][1];

  for (const i of intervals) {
    if (rangeMax < i[1]) {
      count++;
      rangeMax = i[1];
    }
  }

  return count;
}

File: solutions/Java/[1288_Medium] Remove Covered Intervals/removeCoveredIntervals.java
class Solution {
  public void insertionSort(int[][] arr) {
    for (int k = 1; k < arr.length; k++) {
      int[] key = arr[k];
      int i = k - 1;
      while (i >= 0 && (arr[i][0] > key[0] || (arr[i][0] == key[0] && arr[i][1] < key[1]))) {
        arr[i + 1] = arr[i];
        i--;
      }
      arr[i + 1] = key;
    }
  }

  public int removeCoveredIntervals(int[][] intervals) {
    insertionSort(intervals);

    int count = 1;
    int rangeMax = intervals[0][1];

    for (int[] i: intervals) {
      if (rangeMax < i[1]) {
        count++;
        rangeMax = i[1];
      }
    }

    return count;
  }
}
